### DISCUSSION TITLE
Speedy reconstruction with small reconstruction volume, limited angles

### DISCUSSION THREAD

**Discussion by RikVolger** [astra-toolbox/astra-toolbox#363] [Q&A]
Hello!

I'm working with an XRay setup with just 3 angles, where the final reconstruction requires limited resolution (100x100x40 px is fine). The detectors themselves record images of 1548x300 pixels. I'm expecting quite some artefacts still, and that's okay. I'm curious what would be the fastest way to make a lot of reconstructions with this setup? I'm currently trying to GPU-accelerate the SIRT algorithm on an RTX 3080, but I'm getting very limited GPU utilization (<5%).
As I understand it, the ASTRA algorithms generally attempt to split a single reconstruction on that GPU, which doesn't help much in this case. I think a good approach would be to spread multiple reconstructions with the same geometry over the GPU. Does anyone have an idea how to implement this?
Or perhaps another idea to achieve speedy reconstructions with this setup?

Cheers,
Rik

**Comment by adriaangraas** [comment]
Hi Rik,

Thanks for getting in touch! As the reconstruction problem is very small it is quite possible that the individual forward- and backprojection (CUDA kernel) calls that SIRT makes never require the full GPU. Because of the atypical number of angles, there may be ways to speed up the CUDA kernel, however that would require a recompilation of ASTRA, as well as diving into the ASTRA source.

As you said, there are indeed algorithmic ways imaginable to batch-execute the computations for all reconstruction problems (timeframes) simultaneously, using the same geometry. In essence this would run the SIRT algorithm for all the problems in parallel, and there is a good chance that this would yield a speedup. Unfortunately this is not possible with current ASTRA, partially because SIRT always assumes the output to be a single reconstruction volume.

At the moment we are also working on Python software to allow fast alternative implementations for reconstruction algorithms. This may be interesting at some point. However, I expect that the current ASTRA CUDA algorithms are fast enough to execute your small reconstruction problem in seconds, and if that suffices for you then it may be the easiest thing to do. (If the reconstruction is not fast, please let me/us know and we can have a look why that would be the case.)

Best, Adriaan

**Reply by RikVolger** [reply]
Hi Adriaan,

Thanks for the reply, it provides valuable insight!
For now, I'll see how fast I can get things working with the current setup :)

Cheers!

**Comment by askorikov** [comment]
Also, I think GPU utilization with Astra may not be high anyway, because projection/backprojection kernels are memory bandwidth-limited, if I recall correctly. Or is the situation significantly different with more projections as an input or with a higher reconstruction resolution?
